File: generated_code/GeminiPro/Python/046_SplayTree.py
import sys

class SplayTree:
    """
    A complete, self-contained SplayTree class that implements a
    dictionary-like set for integers.

    This implementation includes the essential methods: insert, delete, and search.
    The search operation (and by extension, insert and delete) performs a splay
    operation, moving the accessed node (or its parent if not found) to the root
    to optimize future accesses for that key.
    """

    class _Node:
        """A private node class for the SplayTree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation around the given node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation around the given node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent

        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
            
        if y:
            y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splay operation on node x, moving it to the root.
        This is the core operation of the Splay Tree, using zig, zig-zig,
        and zig-zag rotations.
        """
        while x.parent:
            p = x.parent
            g = p.parent  # Grandparent

            if not g:  # Zig case
                if x == p.left:
                    self._right_rotate(p)
                else:
                    self._left_rotate(p)
            elif x == p.left and p == g.left:  # Zig-Zig (left-left)
                self._right_rotate(g)
                self._right_rotate(p)
            elif x == p.right and p == g.right:  # Zig-Zig (right-right)
                self._left_rotate(g)
                self._left_rotate(p)
            elif x == p.right and p == g.left:  # Zig-Zag (left-right)
                self._left_rotate(p)
                self._right_rotate(g)
            else:  # Zig-Zag (right-left)
                self._right_rotate(p)
                self._left_rotate(g)

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the node with that key is splayed to the root.
        If the key is new, it is inserted and the new node is splayed to the root.

        Args:
            key: The integer key to insert.
        """
        node = self.root
        parent = None
        while node:
            parent = node
            if key == node.key:
                # Key already exists, splay it and we're done
                self._splay(node)
                return
            elif key < node.key:
                node = node.left
            else:
                node = node.right

        new_node = self._Node(key)
        new_node.parent = parent

        if not parent:
            self.root = new_node
        elif key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        
        # Splay the new node to the root
        self._splay(new_node)

    def search(self, key):
        """
        Searches for a key in the tree and splays the accessed node.

        If the key is found, the corresponding node is splayed to the root.
        If the key is not found, the last non-null node accessed during the
        search is splayed to the root.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        node = self.root
        last_node = None
        while node:
            last_node = node
            if key == node.key:
                self._splay(node)
                return True
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        
        # Key not found, splay the last visited node if it exists
        if last_node:
            self._splay(last_node)
            
        return False

    def delete(self, key):
        """
        Deletes a key from the tree.

        The operation first searches for the key, which splays the node (if found)
        or its would-be parent to the root. If the key is found, the node is
        removed, and its two subtrees are joined.

        Args:
            key: The integer key to delete.
        """
        # Search for the key. This splays the found node to the root.
        # If not found, it splays the would-be parent and returns False.
        if not self.search(key):
            return  # Key wasn't in the tree.

        # If search returned True, the node to delete is now the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If there's no left subtree, the right subtree becomes the new tree.
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        else:
            # If a left subtree exists, we find its maximum element, splay it
            # to the root of the left subtree, and then attach the right subtree.
            left_subtree.parent = None
            
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Splay max_node to be the root of the left subtree.
            # We temporarily make the left subtree the main tree for the splay operation.
            self.root = left_subtree
            self._splay(max_node) # self.root is now max_node
            
            # Attach the original right subtree to the new root.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root

File: generated_code/GeminiPro/Java/046_SplayTree.java
/**
 * A Splay Tree implementation that stores a set of integers.
 *
 * A splay tree is a self-balancing binary search tree with the additional
 * property that recently accessed elements are quick to access again.
 * It performs basic operations such as insertion, deletion, and search.
 * After a search or an access, the accessed node is moved to the root of the tree
 * through a series of rotations, a process known as "splaying". This implementation
 * does not allow duplicate keys.
 */
public class SplayTree {

    /**
     * Inner class representing a node in the splay tree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        /**
         * Constructor for a new node.
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty SplayTree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the given node.
     *
     * @param x The node to be rotated.
     * @return The new root of the subtree.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the given node.
     *
     * @param x The node to be rotated.
     * @return The new root of the subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The core splaying operation.
     * This method searches for a key and moves the last accessed node to the root.
     * This is a top-down splaying implementation.
     *
     * @param root The root of the tree (or subtree) to splay.
     * @param key The key to splay around.
     * @return The new root of the splayed tree.
     */
    private Node splay(Node root, int key) {
        if (root == null) {
            return null;
        }

        // Dummy node to simplify linking. header.right points to the left tree (L),
        // and header.left points to the right tree (R).
        Node header = new Node(0);
        Node leftTreeMax = header;
        Node rightTreeMin = header;

        while (true) {
            if (key < root.key) {
                if (root.left == null) break;
                // Zig-Zig case (right rotation)
                if (key < root.left.key) {
                    root = rightRotate(root);
                    if (root.left == null) break;
                }
                // Link the current root to the right tree
                rightTreeMin.left = root;
                rightTreeMin = root;
                root = root.left;
            } else if (key > root.key) {
                if (root.right == null) break;
                // Zig-Zig case (left rotation)
                if (key > root.right.key) {
                    root = leftRotate(root);
                    if (root.right == null) break;
                }
                // Link the current root to the left tree
                leftTreeMax.right = root;
                leftTreeMax = root;
                root = root.right;
            } else { // Found the key
                break;
            }
        }

        // Reassemble the tree
        leftTreeMax.right = root.left;
        rightTreeMin.left = root.right;
        root.left = header.right;
        root.right = header.left;

        return root;
    }

    /**
     * Searches for a key in the tree.
     * If the key is found, it is splayed to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the tree remains unchanged.
     * After insertion, the new node becomes the root of the tree.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // Handle case for an empty tree
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree around the key. The closest node will become the root.
        root = splay(root, key);

        // If the key is already present, do nothing.
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the splay tree.
     * If the key is not in the tree, the tree remains unchanged.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree to bring the node to delete (or its closest value) to the root.
        root = splay(root, key);

        // If the key is not in the tree, do nothing.
        if (root.key != key) {
            return;
        }

        // The node to delete is now the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring its largest element to its root.
            // Splaying for the key 'key' guarantees the max element is the new root
            // because 'key' is larger than any element in the left subtree.
            leftSubtree = splay(leftSubtree, key);

            // The new root of the left subtree has no right child.
            // Attach the original right subtree there.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }
}